### PR TITLE
WIP Synchronise frontlight level with nickel config

### PR DESCRIPTION
**Pull request by dset0x**
A few points about this patch:

* Because I consider nickel an optional external entity I tried to keep it away from core logic. For this reason I [load its value into `settings.reader.lua`](https://github.com/koreader/koreader/compare/master...dset0x:frontlight-sync?expand=1#diff-df44ce4f2932b4d7c52ce1bd4d511defR74) and let the rest of koreader's boot process read _that_ as it's always been doing.

* At the moment, KSM resets the brightness setting when switching from koreader to nickel and vice-versa. I have patches that fix that and once I've cleaned them up I will push them in KSM's way.

---

Issues that make this a WIP for now:

* Currently set to crash when `/mnt/onboard/.kobo/Kobo/Kobo eReader.conf` is missing, but I'll fix once I have the chance.

* The ini parser is very naive, but I could not find a lua library that works with nickel's configuration without breaking it. I'm open to suggestions to making it more robust.

### CODE REVIEW

**Comment by Frenzie** [comment]
Clever!

File: frontend/device/kobo/nickel_conf.lua [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    local correct_section = false
+    local kobo_conf = assert(io.open(kobo_conf_path, "r"))
+    for line in kobo_conf:lines() do
+        if string.match(line, re_AnySection) then
+            correct_section = false
+            if string.match(line, re_PowerOptionsSection) then
+                correct_section = true
+            end
+        end
+        old_intensity = string.match(line, re_FrontLightLevel)
```
Comment (houqp) [inline frontend/device/kobo/nickel_conf.lua]: can we make this local?

File: frontend/device/kobo/nickel_conf.lua [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+        if correct_section and old_intensity then
+            lines[#lines + 1] = string.gsub(line, re_BrightnessValue, new_intensity, 1)
+            remaining_file = kobo_conf:read("*a")
+            break
+        else
+            lines[#lines + 1] = line
+        end
+    end
+    kobo_conf:close()
+
```
Comment (houqp) [inline frontend/device/kobo/nickel_conf.lua]: IIRC, it's possible for kobo config to not have the brightness setting if the user has never set it from the UI. In that case, we should add it manually.

File: frontend/device/kobo/nickel_conf.lua [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    local kobo_conf = assert(io.open(kobo_conf_path, "r"))
+    for line in kobo_conf:lines() do
+        if string.match(line, re_AnySection) then
+            correct_section = false
+            if string.match(line, re_PowerOptionsSection) then
+                correct_section = true
+            end
+        end
+        if correct_section then
+            new_intensity = string.match(line, re_FrontLightLevel)
```
Comment (houqp) [inline frontend/device/kobo/nickel_conf.lua]: declare new_intensity as local outside of the for loop?

File: frontend/ui/uimanager.lua [hunk 4]
```diff
@@ -6,6 +6,7 @@ local Geom = require("ui/geometry")
 local util = require("ffi/util")
 local DEBUG = require("dbg")
 local _ = require("gettext")
+local NickelConf = require("device/kobo/nickel_conf")
```
Comment (houqp) [inline frontend/ui/uimanager.lua]: minor issue. we can put this import into the `Device:isKobo()` if statement.

File: frontend/ui/uimanager.lua [hunk 5]
```diff
@@ -60,8 +61,18 @@ function UIManager:init()
                 self:sendEvent(input_event)
             end
         end
-        if KOBO_LIGHT_ON_START and tonumber(KOBO_LIGHT_ON_START) > -1 then
-            Device:getPowerDevice():setIntensity( math.max( math.min(KOBO_LIGHT_ON_START,100) ,0) )
+        local kobo_light_on_start = tonumber(KOBO_LIGHT_ON_START)
+        if kobo_light_on_start then
+            if kobo_light_on_start >= 0 then
```
Comment (houqp) [inline frontend/ui/uimanager.lua]: same here, could we declare new_intensity as local before this if statement?

**Comment by houqp** [comment]
Looks great! Thanks for cleaning this up :) I added some comments. It would really nice if you can also add some unit tests for the kobo config parsing/updating code.

**Comment by dset0x** [comment]
@houqp Thanks for the comments - as you can see I'm not well versed in lua (yet). I've amended what I could but I've yet to test things well.

I'm all for adding proper tests, but I will be leaving next week and I will be away for quite a long time. Regardless, I will see what I can do until then and let you know.

---

You can follow the relevant KSM discussion [here](http://www.mobileread.com/forums/showthread.php?p=3238503#post3238503)

**Comment by houqp** [comment]
No worry :) It's indeed good quality code. If you don't have time to write the tests, I can do it after it's merged. So are we blocked on KSM's change now?

**Comment by dset0x** [comment]
@houqp It turns out that I had misunderstood some things about KSM that tshering cleared up for me. This patch should work fine regardless of KSM. It is not blocked.

I will let you know by Sunday the latest if I can push unit tests in a reasonable time. Anyway, here's the use-cases that we must definitely catch:

* Simply replace value:
```ini
[OtherThing]
foo=bar
[PowerOptions]
FrontLightLevel=6
[YetAnotherThing]
bar=baz
```

* Create new section and add new line with new value (ie don't be fooled by key of same name) :
```ini
[OtherThing]
FrontLightLevel=6
```

* Insert new value between the two lines (this is extremely important because there was a regression that was replacing "[OtherThing"]) :
```ini
[PowerOptions]
[OtherThing]
```

* If the file is missing, create it and add section and line.

* If the file exists but is empty, add section and line.

**Pull request merged**